// @dreamcomposer/pre/lib/methods/effect-settings.js
/**
 * The default values for the effects.
 */
export const DEFAULT_EFFECT_SETTINGS = Object.freeze({
    fieldOfView: 0.75,
    bloomEnabled: true,
    bloomThreshold: 0.9,
    bloomWeight: 0.05,
    chromaticAberrationEnabled: false,
    chromaticAberrationAmount: 0.0,
    grainEnabled: false,
    grainIntensity: 30,
    imageProcessingBrightness: 1,
    imageProcessingContrast: 1,
    toneMappingEnabled: true,
    vignetteEnabled: true,
    vignetteWeight: 1.5,
    sharpenEnabled: false,
    sharpenAmount: 0.3,
    ditheringEnabled: true,
    ditheringIntensity: 1 / 255,
});
/**
 * Applies some effect settings.
 *
 * @param dcController The DcController object obtained from the init function.
 * @param settings The settings.
 * @returns A promise that resolves when the settings have been applied.
 */
export function setEffectSettings(dcController, settings) {
    return dcController.postMessage({
        channel: "set-effects-settings",
        payload: {
            ...DEFAULT_EFFECT_SETTINGS,
            ...settings,
        },
    });
}
