// @dreamcomposer/pre/lib/pcm/item.js
/**
 * @internal
 */
export var PermissionRole;
(function (PermissionRole) {
    /**Implicit role if the item is no longer available*/
    PermissionRole["NONE"] = "NONE";
    /**Read access to the item data*/
    PermissionRole["USER"] = "USER";
    /**Full access to the item*/
    PermissionRole["OWNER"] = "OWNER";
})(PermissionRole || (PermissionRole = {}));
/**
 * Checks if the given object is a PcmItem.
 *
 * @param obj The object to check.
 * @returns True if the object is a PcmItem, false otherwise.
 */
export function isPcmItem(obj) {
    return (typeof obj === "object" &&
        obj !== null &&
        "id" in obj &&
        "displayName" in obj);
}
/** @internal */
export const mapNativePcmItemToPcmItem = (item) => ({
    id: item.id,
    displayName: item.name,
    description: item.description || undefined,
    flags: item.flags,
    tags: item.tags,
    category: {
        id: item.category.id,
        name: item.category.name,
        description: item.category.description || undefined,
        orderPriority: item.category.order_priority,
    },
    files: item.files.map(mapNativePcmItemFileToPcmItemFile),
    accessRight: item.access_right || PermissionRole.NONE,
    creationDate: new Date(item.creation_date),
});
function mapNativePcmItemFileToPcmItemFile(file) {
    return {
        name: file.name,
        type: file.type,
        description: file.description || undefined,
        extension: file.extension,
        url: file.url,
        creationDate: new Date(file.creation_date),
    };
}
